bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringTable.java

```diff
@@ -0,0 +1,1052 @@
+      MethodDescriptor methodDescriptor, Message message, byte[] bytes, byte[] bytes1, R r)
+      throws Throwable {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public <R extends Message> void batchCoprocessorService(
+      MethodDescriptor methodDescriptor,
+      Message message,
+      byte[] bytes,
+      byte[] bytes1,
+      R r,
+      Callback<R> callback)
+      throws Throwable {
+    throw new UnsupportedOperationException();
+  }
+
+  private boolean checkAndMutateWithSpan(
+      final byte[] row,
+      final byte[] family,
+      final byte[] qualifier,
+      final CompareOp compareOp,
+      final byte[] value,
+      final RowMutations rowMutations)
+      throws IOException {
+    boolean wereMutationsApplied =
+        this.mirroringTracer.spanFactory.wrapPrimaryOperation(
+            new CallableThrowingIOException<Boolean>() {
+              @Override
+              public Boolean call() throws IOException {
+                return MirroringTable.this.primaryTable.checkAndMutate(
+                    row, family, qualifier, compareOp, value, rowMutations);
+              }
+            },
+            HBaseOperation.CHECK_AND_MUTATE);
+
+    if (wereMutationsApplied) {
+      scheduleSequentialWriteOperation(
+          new WriteOperationInfo(rowMutations), this.secondaryAsyncWrapper.mutateRow(rowMutations));
+    }
+    return wereMutationsApplied;
+  }
+
+  @Override
+  public boolean checkAndMutate(
+      byte[] row,
+      byte[] family,
+      byte[] qualifier,
+      CompareOp compareOp,
+      byte[] value,
+      RowMutations rowMutations)
+      throws IOException {
+    try (Scope scope =
+        this.mirroringTracer.spanFactory.operationScope(HBaseOperation.CHECK_AND_MUTATE)) {
+      Log.trace(
+          "[%s] checkAndMutate(row=%s, family=%s, qualifier=%s, compareOp=%s, value=%s, rowMutations=%s)",
+          this.getName(), row, family, qualifier, compareOp, value, rowMutations);
+
+      return checkAndMutateWithSpan(row, family, qualifier, compareOp, value, rowMutations);
+    }
+  }
+
+  @Override
+  public void setOperationTimeout(int i) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int getOperationTimeout() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int getRpcTimeout() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void setRpcTimeout(int i) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int getReadRpcTimeout() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void setReadRpcTimeout(int i) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int getWriteRpcTimeout() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void setWriteRpcTimeout(int i) {
+    throw new UnsupportedOperationException();
+  }
+
```

Comment:
can these be forwarded to the primary?

Reply:
We can, but I think that we should throw UnsupportedOperationException in cases not supported by bigtable client - even when it is the secondary. If users use these methods then they should modify their code and I'd like to inform them about that (by throwing that exception) as early as possible to prevent cases in which users' code works for hbase -> bigtable mirroring, but stops working after reconfiguration to bigtable -> hbase and a rollback and code modification is required.

Reply:
I am ok with fail fast approach here.